scripts/mimc_constraints.sage

```diff
@@ -63,26 +63,47 @@ def output_valid_configs_and_constraints(r):
     # output_valid_config_and_constraints(r, log_2_r, 23)
 
 
+def check_polynomial(r, e):
+    """
+    Check that the polynomial representing a mimc round is not sparse (in fact,
+    it follows immediately from the binomial theorem that no coefficients can
+    be zero for r prime, so this is essentially a sanity check).
```

Comment:
Right.  I missed the restriction  r > e > 2.  Thanks.

I wasn't referring to anything in the specs specifically.  My reasoning was that each coefficient can be written as a product of integers all less than e, so cannot be 0 mod r for e < r ... but perhaps I'm overlooking something.

Should be fixed now, but let me know if I got something wrong.  (I added this after our discussion, but given that (AFAICT) it holds by definition, I'm also happy to just remove it from the script.)

Reply:
> Right. I missed the restriction r > e > 2. Thanks.

Perfect thanks (for completeness, though it goes without saying really -> other restriction for dense polys is to **not** exponentiate a monomial in the first place -> i.e. if the key and round constant happen to be additive inverses, we reduce `(k + ci + x)^e` to `x^e`)

Reply:
However this dummy/special case does not seem to be caught by your script AFAICT since you model `mimc_poly` as a bi-variate polynomial in `x` and `k`. The mimc polynomials (i.e. round polynomial or resulting polynomial after n rounds) are univariate (the key and round constants are fixed) though

Reply:
> for completeness, though it goes without saying really -> other restriction for dense polys is to not exponentiate a monomial in the first place -> i.e. if the key and round constant happen to be additive inverses, we reduce (k + ci + x)^e to x^e

Indeed.  Are you suggesting this should be in the comments?

> However this dummy/special case does not seem to be caught by your script AFAICT since you model mimc_poly as a bi-variate polynomial in x and k. The mimc polynomials (i.e. round polynomial or resulting polynomial after n rounds) are univariate (the key and round constants are fixed) though. -> You'd need to have a partial evaluation of your bi-variate poly on k to exhibit this and recover the sparse poly where all powers of k are 0 except k^0 which gives the monomial term with max power for x

In a given invocation, `k` is indeed fixed.  In fact, if `k` represents `key + rc`, it is fixed for any given round, as `x` is.
If `k` happens to be 0 for some round, the result will indeed be `x^e` - this is an unavoidable result of the mimc definition.  But since we don't know `k` in advance, I'm not entirely sure how the script could "catch" any given case.  Are you suggesting that we choose some specific value of `k`?  Would that imply something further?  

I'll remove this check for now as it seems to be creating more confusion than value (my mistake for adding it without fully understanding the intention).  If we identify something concrete to be checked empirically (i.e. something that does not follow from the restrictions on `r`, `e`, `rounds` etc) then let's revisit and maybe add it in another PR (concrete checks for specific chosen values may be more suitable in a separate script anyway, or in the zeth-specs).

Reply:
> Are you suggesting that we choose some specific value of k? Would that imply something further?

I'm just chiming in here to make sure we define strict boundaries around the claims we make (make sure we define for which constraints on `r`, `e` etc, the claim on polynomial density applies etc.).
If we just unroll a bi-variate poly (and thus "unroll" the binomial Th) we clearly see that we have all terms, but these are just variables and do not convey much information w.r.t the **univariate** polynomials we will manipulate with MiMC -> these may well vanish when choosing the concrete parameters leading to a concrete sparse polynomial in `x` (namely, if the second variable `k = 0`). Again, simply chiming in here to make sure we're clear on that (and to make sure all of this is clearly defined/written in the related code to define the boundaries in which the claim is true).
Of course, in practice, we select the round constants, but can't really control "how they could interfere with the key" since we don't manually select the key ourselves (but rather use the neighboring leaf in the Merkle three as key), yet we should remember this "very unlikely" edge case IMO (unlikely due to the uniqueness of the additive inverse) as it remains a valid counter-example for the initial claim.

Reply:
So this is just to say: that perfectly fine to keep this function/sanity check as long as we constrain the variables (e, r etc) well enough and document well the intent/edge cases. In fact, the check carries interesting information since it intrinsically constrains the value of the exponent (if you have, e.g. `r = 7` and `(x + k)^8` -> the asserts will be thrown because the coefficients will wrap around 7 and vanish) - the TL;DR: I was simply confused by the initial formulation of the comment :)

Reply:
As discussed offline, we'll keep this in the specs (where the params are selected).